(FM-7955) Update to use Transport helper code

da-ar opened the pull request:
Tasks written for remote transports can be written as `def task(params, remote)`  to expose the helper code.

The `remote` parameter provides methods for accessing configuration of the target as well as access to the `transport` object via `remote.transport`.

Example use: https://github.com/puppetlabs/cisco_ios/pull/335

da-ar commented:
Working on fixing up the code causing the test failures.

da-ar commented:
Updated location for the example use: https://github.com/da-ar/cisco_ios/commit/68320c1d7a6ea33759db5a464d7a877b3132d959

davidmalloncares commented:
Hey @lucywyman would you have a few minutes to review this for us please?

lucywyman commented on spec/helper_spec.rb:
```diff
@@ -96,3 +106,27 @@ def task(params)
+    { "protocol": 'remote', "remote-transport": 'wibble' }
+  end
+  let(:input) do
+    { 'name': 'Lucy', '_target': target }
```
Haha :)

lucywyman commented on files/task_helper.rb:
```diff
@@ -18,7 +18,7 @@ def to_h
   end
 
-  def task(params = {})
+  def task(params = {}, remote = nil)
```
Does this need to be it's own argument to the task helper, or can it just be a keyword argument that's passed in when needed?

da-ar replied:
I'm happy to change it if you have a preference, could you give me an example of how you would like this to look?

adreyer replied:
Yeah this won't work long term when we may have other optional data that is important to the task implementation. 

I think making `remote` an instance method is probably a better API

lucywyman reviewed (commented):
@da-ar This works for me! Would you mind updating the documentation and adding an example of using it?

da-ar commented:
@lucywyman Hey, I've finally updated the docs. Sorry for the delay.

lucywyman commented:
Awesome! I'm going to play with this a bit more on Monday morning, but overall looks good. Thanks!

adreyer commented on files/task_helper.rb:
```diff
@@ -46,7 +46,13 @@ def self.run
     # @return [Hash] The result of the task
-    result = new.task(params)
+
+    result = if params.dig(:_target, :protocol) &&
```
the protocol of a remote target is not necessarily `remote`.  the `_target` metaparam exists solely for remote targets which may use other protocols.

`dig` and `==` is redundant too

adreyer commented on files/task_helper.rb:
```diff
@@ -46,7 +46,13 @@ def self.run
+
+    result = if params.dig(:_target, :protocol) &&
+                params[:_target][:protocol] == 'remote'
+               new.task(params, RemoteContext.new(params))
```
Why not just put this in initialize?

da-ar committed:
(FM-7955) Expose Remote Context to tasks

Tasks written for remote transports can access the `context` object which exposes the transport to the task via `context.transport`.

adreyer commented on files/task_helper.rb:
```diff
@@ -46,7 +48,10 @@ def self.run
-    result = new.task(params)
+
+    tsk = new
+    tsk.context = RemoteContext.new(params) if params.key? :_target
```
why not move this logic to the initialize method?

da-ar replied:
Because `RemoteContext` need the params, which are read by this function
https://github.com/puppetlabs/puppetlabs-ruby_task_helper/blob/6c5b8c51f6f233a78587d9f80313057a31a85be1/files/task_helper.rb#L43-L46

This looked like the least ugly way of doing that

adreyer commented on spec/helper_spec.rb:
```diff
@@ -96,3 +106,27 @@ def task(params)
+
+  it 'runs an remote task' do
+    allow(STDIN).to receive(:read).and_return(input.to_json)
+    allow(Puppet::ResourceApi::Transport).to receive(:connect)
```
I thikn we should have a real mock transport for this test probably.

davidmalloncares commented:
@adreyer would it be possible to get this merged as a first stage? I'd like to get the initial work over the line as we currently don't have the resource to work on your proposed updates right now

da-ar commented:
closing due to lack of interest
